File: flutter/lib/models/telemetry.dart
class Telemetry {
  final int id;
  final int seq;
  final int ts;
  final DateTime receivedAt;
  final int rssi;
  final double snr;

  // GPS
  final double? lat;
  final double? lon;
  final double? alt;
  final double? speedKmh;
  final int satellites;
  final bool gpsFix;

  // BME280
  final double bmeTemp;
  final double bmeHum;
  final double bmePres;

  // MS5611
  final double msTemp;
  final double msPres;
  final double msAlt;

  // BNO085
  final double roll;
  final double pitch;
  final double yaw;

  // INA219
  final double voltage;
  final double currentMa;
  final double powerMw;

  const Telemetry({
    required this.id,
    required this.seq,
    required this.ts,
    required this.receivedAt,
    required this.rssi,
    required this.snr,
    this.lat,
    this.lon,
    this.alt,
    this.speedKmh,
    required this.satellites,
    required this.gpsFix,
    required this.bmeTemp,
    required this.bmeHum,
    required this.bmePres,
    required this.msTemp,
    required this.msPres,
    required this.msAlt,
    required this.roll,
    required this.pitch,
    required this.yaw,
    required this.voltage,
    required this.currentMa,
    required this.powerMw,
  });

  factory Telemetry.fromJson(Map<String, dynamic> j) => Telemetry(
        id: j['id'] ?? 0,
        seq: j['seq'] ?? 0,
        ts: j['ts'] ?? 0,
        receivedAt: DateTime.tryParse(j['received_at'] ?? '') ?? DateTime.now(),
        rssi: j['rssi'] ?? 0,
        snr: (j['snr'] ?? 0).toDouble(),
        lat: (j['lat'] as num?)?.toDouble(),
        lon: (j['lon'] as num?)?.toDouble(),
        alt: (j['alt'] as num?)?.toDouble(),
        speedKmh: (j['speed_kmh'] as num?)?.toDouble(),
        satellites: j['satellites'] ?? 0,
        gpsFix: j['gps_fix'] ?? false,
        bmeTemp: (j['bme_temp'] ?? 0).toDouble(),
        bmeHum: (j['bme_hum'] ?? 0).toDouble(),
        bmePres: (j['bme_pres'] ?? 0).toDouble(),
        msTemp: (j['ms_temp'] ?? 0).toDouble(),
        msPres: (j['ms_pres'] ?? 0).toDouble(),
        msAlt: (j['ms_alt'] ?? 0).toDouble(),
        roll: (j['roll'] ?? 0).toDouble(),
        pitch: (j['pitch'] ?? 0).toDouble(),
        yaw: (j['yaw'] ?? 0).toDouble(),
        voltage: (j['voltage'] ?? 0).toDouble(),
        currentMa: (j['current_ma'] ?? 0).toDouble(),
        powerMw: (j['power_mw'] ?? 0).toDouble(),
      );
}

File: backend/routers/telemetry.py
from fastapi import APIRouter, Depends, HTTPException, Header, status
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, desc
from typing import Optional

from database import get_db
from models import TelemetryRecord
from schemas import TelemetryIn, TelemetryOut
from config import settings
from routers.ws import broadcast

router = APIRouter(tags=["telemetry"])


def _verify_key(x_api_key: str = Header(...)):
    if x_api_key != settings.api_key:
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Bad API key")


@router.post("/telemetry", status_code=status.HTTP_201_CREATED, response_model=TelemetryOut)
async def ingest(
    body: TelemetryIn,
    db: AsyncSession = Depends(get_db),
    _: None = Depends(_verify_key),
):
    record = TelemetryRecord(
        seq=body.seq,
        ts=body.ts,
        rssi=body.rssi,
        snr=body.snr,
        lat=body.gps.lat,
        lon=body.gps.lon,
        alt=body.gps.alt,
        speed_kmh=body.gps.spd,
        heading=body.gps.hdg,
        satellites=body.gps.sat,
        gps_fix=body.gps.fix,
        bme_temp=body.bme.temperature_c,
        bme_hum=body.bme.humidity_pct,
        bme_pres=body.bme.pressure_hpa,
        ms_temp=body.ms.temperature_c,
        ms_pres=body.ms.pressure_hpa,
        ms_alt=body.ms.altitude_m,
        roll=body.imu.roll_deg,
        pitch=body.imu.pitch_deg,
        yaw=body.imu.yaw_deg,
        voltage=body.pwr.voltage_v,
        current_ma=body.pwr.current_ma,
        power_mw=body.pwr.power_mw,
    )
    db.add(record)
    await db.commit()
    await db.refresh(record)

    out = TelemetryOut.model_validate(record)
    await broadcast(out.model_dump(mode="json"))
    return out


@router.get("/telemetry", response_model=list[TelemetryOut])
async def list_telemetry(
    limit: int = 100,
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(TelemetryRecord).order_by(desc(TelemetryRecord.id)).limit(limit)
    )
    return result.scalars().all()


@router.get("/telemetry/latest", response_model=Optional[TelemetryOut])
async def latest(db: AsyncSession = Depends(get_db)):
    result = await db.execute(
        select(TelemetryRecord).order_by(desc(TelemetryRecord.id)).limit(1)
    )
    return result.scalar_one_or_none()
